prov/util/src/util_cntr.c

```diff
@@ -0,0 +1,335 @@
+			return -FI_EINVAL;
+		}
+		/* fall through */
+	case FI_WAIT_UNSPEC:
+	case FI_WAIT_FD:
+		break;
+	default:
+		FI_WARN(prov, FI_LOG_CNTR, "unsupported wait object\n");
+		return -FI_EINVAL;
+	}
+
+	return 0;
+}
+
+static uint64_t ofi_cntr_read(struct fid_cntr *cntr_fid)
+{
+	struct util_cntr *cntr = container_of(cntr_fid, struct util_cntr, cntr_fid);
+	assert(cntr->cntr_fid.fid.fclass == FI_CLASS_CNTR);
+
+	cntr->progress(cntr);
+
+	return ofi_atomic_get64(&cntr->cnt);
+}
+
+static uint64_t ofi_cntr_readerr(struct fid_cntr *cntr_fid)
+{
+	struct util_cntr *cntr = container_of(cntr_fid, struct util_cntr, cntr_fid);
+	assert(cntr->cntr_fid.fid.fclass == FI_CLASS_CNTR);
+
+	cntr->progress(cntr);
+
+	return ofi_atomic_get64(&cntr->err);
+}
+
+static int ofi_cntr_add(struct fid_cntr *cntr_fid, uint64_t value)
+{
+	struct util_cntr *cntr = container_of(cntr_fid, struct util_cntr, cntr_fid);
+	assert(cntr->cntr_fid.fid.fclass == FI_CLASS_CNTR);
+
+	ofi_atomic_add64(&cntr->cnt, value);
+
+	if(cntr->wait)
+		cntr->wait->signal(cntr->wait);
+
+	return FI_SUCCESS;
+}
+
+static int ofi_cntr_adderr(struct fid_cntr *cntr_fid, uint64_t value)
+{
+	struct util_cntr *cntr = container_of(cntr_fid, struct util_cntr, cntr_fid);
+	assert(cntr->cntr_fid.fid.fclass == FI_CLASS_CNTR);
+
+	ofi_atomic_add64(&cntr->err, value);
+
+	if(cntr->wait)
+		cntr->wait->signal(cntr->wait);
+
+	return FI_SUCCESS;
+}
+
+static int ofi_cntr_set(struct fid_cntr *cntr_fid, uint64_t value)
+{
+	struct util_cntr *cntr = container_of(cntr_fid, struct util_cntr, cntr_fid);
+	assert(cntr->cntr_fid.fid.fclass == FI_CLASS_CNTR);
+
+	ofi_atomic_initialize64(&cntr->cnt, value);
+
+	if(cntr->wait)
+		cntr->wait->signal(cntr->wait);
+
+	return FI_SUCCESS;
+}
+
+static int ofi_cntr_seterr(struct fid_cntr *cntr_fid, uint64_t value)
+{
+	struct util_cntr *cntr = container_of(cntr_fid, struct util_cntr, cntr_fid);
+	assert(cntr->cntr_fid.fid.fclass == FI_CLASS_CNTR);
+
+	ofi_atomic_initialize64(&cntr->err, value);
+
+	if(cntr->wait)
+		cntr->wait->signal(cntr->wait);
+
+	return FI_SUCCESS;
+}
+
+static int ofi_cntr_wait(struct fid_cntr *cntr_fid, uint64_t threshold, int timeout)
+{
+	struct util_cntr *cntr = container_of(cntr_fid, struct util_cntr, cntr_fid);
+	uint64_t current_ms;
+	uint64_t finish_ms;
+	uint64_t err = ofi_cntr_readerr(cntr_fid);
+
+	assert(cntr->cntr_fid.fid.fclass == FI_CLASS_CNTR);
+
+	if (threshold <= ofi_cntr_read(cntr_fid))
+		return FI_SUCCESS;
+
+	assert(cntr->wait);
+
+	current_ms = fi_gettime_ms();
+	finish_ms = (timeout < 0) ? UINT64_MAX : current_ms + timeout;
+	for (; timeout < 0 || current_ms < finish_ms;
+	    current_ms = fi_gettime_ms()) {
+		timeout = timeout < 0 ? timeout : (int)(finish_ms - current_ms);
+		fi_wait(&cntr->wait->wait_fid, timeout);
+		cntr->progress(cntr);
+		if (threshold <= ofi_cntr_read(cntr_fid))
+			return FI_SUCCESS;
+		else if (err != ofi_cntr_readerr(cntr_fid))
+			return -FI_EAVAIL;
+	}
```

Comment:
Calling ofi_cntr_read*() results in multiple calls to progress().  I would slightly rework the code to replace all ofi_cntr_read*() calls with ofi_atomic_get64().  An additional call to progress() would then be needed at the location indicated above.